ext/intl/locale/locale_methods.c

```diff
@@ -1238,7 +1238,7 @@ PHP_FUNCTION(locale_filter_matches)
 		if( token && (token==cur_lang_tag) ){
 			/* check if the char. after match is SEPARATOR */
 			chrcheck = token + (strlen(cur_loc_range));
-			if( isIDSeparator(*chrcheck) || isEndOfTag(*chrcheck) ){
+			if( isIDSeparator(*chrcheck) || isKeywordSeparator(*chrcheck) || isEndOfTag(*chrcheck) ){
```

Comment:
This code checks two language tags to see if the first tag is a prefix of the second tag. It does a "look-ahead" here to see if the next character is a "separator" character. If the next character isn't a separator character, then we can't treat the first tag as a prefix, so we'll return false. If it is a separator, then we can reliably assume it is a valid language tag prefix for the longer canonicalized form.

Our previous logic treated only `_`, `-`, and `\0` as separator characters, which worked well for the way tags were canonicalized before ICU 70.1. For example:

1. Given the language tags `sl_IT_nedis-a-kirti-x-xyz` and `sl_IT_Nedis`
2. These were canonicalized to `sl_IT_NEDIS_A_KIRTI_X_XYZ` and `sl_IT_NEDIS`
3. We lowercased both strings to `sl_it_nedis_a_kirti_x_xyz` and `sl_it_nedis`
4. Then we checked whether `sl_it_nedis` is a prefix of `sl_it_nedis_a_kirti_x_xyz`
5. It's a prefix and the next character in the `cur_loc_range` is a separator
6. So, we return `true`

The canonicalization has changed, though, so `sl_IT_nedis-a-kirti-x-xyz` is now canonicalized to `sl_IT_NEDIS@a=kirti;x=xyz`. Converting them to lowercase for comparison, we now have `sl_it_nedis@a=kirti;x=xyz` and `sl_it_nedis`. Certainly, `sl_it_nedis` is a prefix of `sl_it_nedis@a=kirti;x=xyz`, but since the next character in the later is an @-symbol (and not one of the separators we're checking for), we assume that `sl_it_nedis` is not a valid prefix.

It turns out that the @-symbol is a keyword separator, and we already have `isKeywordSeparator()` to check for it. Adding this to the `if` statement here ensures that `locale_filter_matches()` works for the new canonicalized forms.

Reply:
Thanks for the detailed explanation!